### ISSUE TITLE
[explicit-function-return-type]: unexpected error for react component class methods

### ISSUE THREAD

**Issue by publicJorn** [typescript-eslint/typescript-eslint#1404]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": ["error"]
  }
}
```

> The default rule, or any option set

```TSX
import React, { Component } from 'react'

class MyComponent extends Component<{}, {}> {
  componentDidMount() {
    // do something
  }

  render() {
    return <p>test</p>
  }
}
```

**Expected Result**

No warnings or errors on react class methods.  

**Actual Result**

Getting error "Missing return type on function." on `componentDitMount` and `render` and other Component class methods.

**Additional Info**

I expect no error, because the interface in `@types/react` extends [componentLifeCycle](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L564), which contains the declarations for these methods.

Is eslint to eager, or unable to parse the interface? Or am I doing it wrong?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.13.0` |
| `@typescript-eslint/parser`        | `2.13.0` |
| `TypeScript`                       | `3.7.4` |
| `ESLint`                           | `6.5.1` (react-scripts 3.2.0) | 
| `node`                             | `10.16.0` |
| `npm`                              | `6.13.4` |

**Comment by publicJorn** [comment]
Extra info; My `eslintrc.js`:

```js
module.exports = {
  parser: '@typescript-eslint/parser',
  extends: [
    'standard',
    'standard-react',
    'plugin:react/recommended',
    'plugin:@typescript-eslint/recommended',
    'prettier/@typescript-eslint',
    'plugin:prettier/recommended',
  ],
  plugins: [
    'react',
    '@typescript-eslint',
    'prettier',
  ],
  env: {
    browser: true,
    jest: true
  },
  
  rules: {
    'prettier/prettier': ['error', {
      tabWidth: 2,
      useTabs: false,
      semi: false,
      singleQuote: true,
      trailingComma: 'all',
      arrowParens: 'always'
    }],
  }
}
```

**Comment by bradzacher** [comment]
This is working as intended.

There is no return type annotation on those methods.

The rule does not use type information, so it does not and cannot interrogate the types of any interfaces or super classes to see what functions are defined.

Regardless, the rule is about _clearly_ creating contracts for your methods/functions, so interface/super class methods are not something the rule should check, because they are rarely colocated with the method or easy to find, so they don't serve as clear contracts.

**Comment by publicJorn** [comment]
> so interface/super class methods are not something the rule should check

Given your explanation I understand. It would have saved me a ton of typing though..

Thanks for your reply!

**Comment by sombhotlu** [comment]
My .eslintrc

```javascript
{
    "parserOptions": {
        "ecmaVersion": 2019,
        "sourceType": "module",
        "ecmaFeatures": {
            "jsx": true
        }
    },
    "extends": ["eslint:recommended", "eslint-config-prettier"],
    "rules": {
        "strict": ["error", "never"]
    },
    "env": {
        "browser": true
    },
    "overrides": [
        {
            "files": "**/*.+(ts|tsx)",
            "parser": "@typescript-eslint/parser",
            "parserOptions": {
                "project": "./tsconfig.json"
            },
            "plugins": ["@typescript-eslint/eslint-plugin"],
            "extends": [
                "plugin:react/recommended",
                "plugin:@typescript-eslint/eslint-recommended",
                "plugin:@typescript-eslint/recommended",
                "eslint-config-prettier/@typescript-eslint"
            ],
            "rules": {
                "@typescript-eslint/interface-name-prefix": [
                    "error",
                    {
                        "prefixWithI": "always"
                    }
                ]
            }
        }
    ]
}
```

**Expected Result**

No warnings or errors on react class methods.

**Actual Result**

Getting error "Missing return type on function." on componentDitMount and render and other Component class methods.

What to do to avoid adding types to lifecycle methods of react ?

**Comment by bradzacher** [comment]
You cannot. As per the comment I closed this issue with.
